### ISSUE TITLE
database is locked

### ISSUE THREAD

**Issue by aixiangwang** [tensorflow/federated#2999]
When I try to increase the train_clients_per_round ,there is an error is that database is locked.I then individually tested the clients with each error and found that they were accessible.
![image](https://user-images.githubusercontent.com/71694385/175763072-e581ac92-7008-42f3-9313-97350f5ee4d6.png)

**Comment by zcharles8** [comment]
Hi @aixiangwang. Can you provide the information requested on the new bug template, including:

* OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
* Python package versions (e.g., TensorFlow Federated, TensorFlow)
* A minimal reproduction of the bug.

This looks like some edge case around the SQL-backed datasets TFF provides, but without the information above I'm not certain what's actually going on.

**Comment by aixiangwang** [comment]
Thank you for your reply.I use tensorflow2.8 and tensorflow-federated0.21 running on the Windows 10.
![image](https://user-images.githubusercontent.com/71694385/176070349-8fb05f82-b599-4246-a2d6-8e26c7857b01.png)
![image](https://user-images.githubusercontent.com/71694385/176070597-ceeffbb3-8235-4f3f-8aa0-c4a782a5da16.png)
you can see client_id = 'f3928_39' in the error appear in lastest 10 randomly selected clients,that make the error.But I found client_id = 'f3928_39' have happened several times before.So I'm very confused.Looking forward to your reply!

**Comment by aixiangwang** [comment]
> Hi @aixiangwang. Can you provide the information requested on the new bug template, including:
> 
> * OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
> * Python package versions (e.g., TensorFlow Federated, TensorFlow)
> * A minimal reproduction of the bug.
> 
> This looks like some edge case around the SQL-backed datasets TFF provides, but without the information above I'm not certain what's actually going on.

Thank you for your reply.I use tensorflow2.8 and tensorflow-federated0.21 running on the Windows 10.
![image](https://user-images.githubusercontent.com/71694385/176070349-8fb05f82-b599-4246-a2d6-8e26c7857b01.png)
![image](https://user-images.githubusercontent.com/71694385/176070597-ceeffbb3-8235-4f3f-8aa0-c4a782a5da16.png)
you can see client_id = 'f3928_39' in the error appear in lastest 10 randomly selected clients,that make the error.But I found client_id = 'f3928_39' have happened several times before.So I'm very confused.Looking forward to your reply!

**Comment by zcharles8** [comment]
Could you add the full code that actually causes this bug? Even better, if you can narrow it down to a smaller reproduction of it, that'd be really helpful.

**Comment by aixiangwang** [comment]
The complete code example is /tensorflow_federated/ simple_fedavg in the current Github project path.I just increased the train_clients_per_round parameter in emnist_fedavg_main.py from 2 to 10, which represents the number of clients sampled per round.
![image](https://user-images.githubusercontent.com/71694385/182100050-e7bc304b-c685-4f2e-9b82-81018b4ad465.png)

Some of the key code in the example is shown below：
train_data, test_data = get_emnist_dataset()

  def tff_model_fn():
    """Constructs a fully initialized model for use in federated averaging."""
    keras_model = create_original_fedavg_cnn_model(only_digits=True)
    loss = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
    metrics = [tf.keras.metrics.SparseCategoricalAccuracy()]
    return tff.learning.from_keras_model(
        keras_model,
        loss=loss,
        metrics=metrics,
        input_spec=train_data.element_type_structure)

  iterative_process = simple_fedavg_tff.build_federated_averaging_process(
      tff_model_fn, server_optimizer_fn, client_optimizer_fn)
  server_state = iterative_process.initialize()
  # Keras model that represents the global model we'll evaluate test data on.
  keras_model = create_original_fedavg_cnn_model(only_digits=True)
  for round_num in range(FLAGS.total_rounds):
    sampled_clients = np.random.choice(
        train_data.client_ids,
        size=FLAGS.train_clients_per_round,
        replace=False)
    print(sampled_clients)
    sampled_train_data = [
        train_data.create_tf_dataset_for_client(client)
        for client in sampled_clients
    ]
    server_state, train_metrics = iterative_process.next(
        server_state, sampled_train_data)
    print(f'Round {round_num}')
    print(f'\tTraining metrics: {train_metrics}')
    if round_num % FLAGS.rounds_per_eval == 0:
      server_state.model.assign_weights_to(keras_model)
      accuracy = evaluate(keras_model, test_data)
      print(f'\tValidation accuracy: {accuracy * 100.0:.2f}%')

As shown in the figure below, the dataset loaded successfully and went through several iterations successfully.
![image](https://user-images.githubusercontent.com/71694385/182101160-d2a0bb5d-d2f7-4522-a75a-93c14120541c.png)
However, an error occurs at a later turn, as shown in the figure because the database is locked and the next federated procedure fails.
![image](https://user-images.githubusercontent.com/71694385/182101927-5bb080e0-2153-4570-ae0a-f94212d0509c.png)
![image](https://user-images.githubusercontent.com/71694385/182102009-4835fb4d-6c83-489f-ae41-c2b07ec6a7e4.png)
Looking forward to your reply!
